### PR TITLE
graph-builder/registry: concurrent tag processing

### PR DESCRIPTION
**Pull request by steveej**
This introduces concurrent processing of tags to potentially reduce
scrape time significantly. This causes the resulting release vector to
be ordered non-deterministically, which had to be considered in the
tests.

Other notable changes:
* Standardize log output throughout all operations when processing a tag
* Introduce graph_builder::registry::cache for convenience

---

TODO
* [x] Depends on https://github.com/camallo/dkregistry-rs/pull/120
* [x] Depends on #195 
* [x] Depends on #206 
* [x] Includes #207 
* [x] minimize critical section for cache locking
* [x] allow unordered processing of the tags
* [x] Make parallelization configurable?

### CODE REVIEW

**Comment by lucab** [comment]
> This involved switching from tokio's current_thread::Runtime to its multi-threaded runtime

I'm not sure I follow the reasoning behind this. Why was concurrency (i.e. buffering futures) not enough and you had to introduce parallelism at the runtime level?

**Comment by steveej** [comment]
> I'm not sure I follow the reasoning behind this. Why was concurrency (i.e. buffering futures) not enough and you had to introduce parallelism at the runtime level?

I didn't have to switch the runtime, but I wanted to. It enables true parallelism which is more than asynchronous I/O.

**Comment by steveej** [comment]
[test]

**Comment by steveej** [comment]
@vrutkovs I think something's broken with the CI setup here.

/retest

**Comment by vrutkovs** [comment]
```
Could not retrieve mirrorlist http://mirrorlist.centos.org/?release=7&arch=x86_64&repo=os&infra=container error was
12: Timeout on http://mirrorlist.centos.org/?release=7&arc... Less than 1000 bytes/sec transferred the last 30 seconds')
```
in images

/retest

**Comment by steveej** [comment]
Another network related failure it seems.

/retest

File: graph-builder/src/registry.rs
Comment (LalatenduMohanty) [inline graph-builder/src/registry.rs]: Can we add a Jira card for this and remove the TODO here?

**Review by LalatenduMohanty — CHANGES_REQUESTED** [review]

File: graph-builder/src/config/options.rs
Comment (vrutkovs) [inline graph-builder/src/config/options.rs:80]: Lets update a configmap in `dist/openshift/cincinnati.yaml` with the default value so that it would not be hidden in the source code

File: graph-builder/src/graph.rs
Comment (vrutkovs) [inline graph-builder/src/graph.rs:167]: Not sure why lifetime is changed here

File: graph-builder/src/registry.rs
Comment (vrutkovs) [inline graph-builder/src/registry.rs:57]: I'm a bit worried about cache growth - do we know if its significant? Should it be periodically purged? Should we use some kind of LRU here?

File: graph-builder/src/registry.rs
Comment (vrutkovs) [inline graph-builder/src/registry.rs:305]: TODO: double `metadata` is confusing, this should be reworked (in a follow up PR)?

File: graph-builder/tests/e2e/mod.rs
Comment (vrutkovs) [inline graph-builder/tests/e2e/mod.rs:71]: Does it have to be defined in `for_each`? Could it be defined once instead?

File: graph-builder/tests/net/quay_io/mod.rs
Comment (vrutkovs) [inline graph-builder/tests/net/quay_io/mod.rs]: Lets use a `static` for this magic number (preferably defined in `graph-builder/src/config/options.rs`)

File: graph-builder/src/graph.rs
Comment (steveej) [reply graph-builder/src/graph.rs:167]: This was part of the compiler warning fix. I noticed it while working on this file.

File: graph-builder/src/registry.rs
Comment (steveej) [reply graph-builder/src/registry.rs:57]: I think the discussion isn't in the scope of this PR, as the cache semantics aren't changed here, but only how the cache is accessed.

File: graph-builder/src/registry.rs
Comment (steveej) [reply graph-builder/src/registry.rs:305]: Not in scope for this PR, as the code you're commenting on is not actually new but just changed its position and indentation.

File: graph-builder/tests/e2e/mod.rs
Comment (steveej) [reply graph-builder/tests/e2e/mod.rs:71]: The macro makes direct use of the variable `edge` which is local to this `for_each`. I also don't foresee the usage of the macro outside of this location.

File: graph-builder/src/registry.rs
Comment (vrutkovs) [reply graph-builder/src/registry.rs:57]: Agree - lets watch the memory consumption on stage and act if it keeps growing

**Comment by steveej** [comment]
Thanks for the review @LalatenduMohanty and @vrutkovs. I addressed all the comments, and, notably,  changed the default concurrency to 16.

**Comment by vrutkovs** [comment]
LGTM! I'll merge this once e2e tests pass

File: dist/openshift/cincinnati.yaml
Comment (vrutkovs) [inline dist/openshift/cincinnati.yaml]:
```suggestion
    value: "16"
```

This causes
`
 error: unable to load template file "dist/openshift/cincinnati.yaml": unable to decode "dist/openshift/cincinnati.yaml": v1.Template.Parameters: []v1.Parameter: v1.Parameter.Value: ReadString: expects " or n, but found 1, error found in #10 byte of ...|,"value":16}]}|..., bigger context ...|","name":"GB_UPSTREAM_FETCH_CONCURRENCY","value":16}]}|...
`
in e2e tests

**Review by vrutkovs — CHANGES_REQUESTED** [review]

**Comment by vrutkovs** [comment]
/hold
template is broken

File: dist/openshift/cincinnati.yaml
Comment (steveej) [reply dist/openshift/cincinnati.yaml]: Oh, thanks for catching this.

**Comment by steveej** [comment]
>  template is broken

Should be fixed now, thanks for catching!

**Comment by vrutkovs** [comment]
graph-builder [fails to start](https://storage.googleapis.com/origin-ci-test/pr-logs/pull/openshift_cincinnati/174/pull-ci-openshift-cincinnati-master-e2e/89/artifacts/e2e/pods/cincinnati-e2e_cincinnati-1-lx7zz_cincinnati-graph-builder.log):
```
error: Invalid value for '--upstream.registry.fetch_concurrency <fetch_concurrency>': invalid digit found in string
```

**Comment by steveej** [comment]
> `error: Invalid value for '--upstream.registry.fetch_concurrency <fetch_concurrency>': invalid digit found in string`

Does this mean that the concurrency value is `<fetch_concurrency>`, which means that there's still a mistake in the template or how it's evaluated.

**Comment by steveej** [comment]
> > `error: Invalid value for '--upstream.registry.fetch_concurrency <fetch_concurrency>': invalid digit found in string`
> 
> Does this mean that the concurrency value is `<fetch_concurrency>`, which means that there's still a mistake in the template or how it's evaluated.

I think I fixed it with this patch which uses the correct braces:

```diff
diff --git a/dist/openshift/cincinnati.yaml b/dist/openshift/cincinnati.yaml
index 2b87a87..0d6cb0d 100644
--- a/dist/openshift/cincinnati.yaml
+++ b/dist/openshift/cincinnati.yaml
@@ -70,7 +70,7 @@ objects:
                 "--upstream.registry.url", "$(REGISTRY)",
                 "--upstream.registry.repository", "$(REPOSITORY)",
                 "--upstream.registry.credentials_path", "/etc/secrets/registry-credentials",
-                "--upstream.registry.fetch_concurrency", "$(GB_UPSTREAM_FETCH_CONCURRENCY)",
+                "--upstream.registry.fetch_concurrency", "${GB_UPSTREAM_FETCH_CONCURRENCY}",
               ]
               ports:
                 - name: graph-builder
```

**Comment by vrutkovs** [comment]
e2e passes, great job!

/hold cancel

**Review by vrutkovs — APPROVED** [review]
/lgtm

**Comment by LalatenduMohanty** [comment]
> Thanks for the review @LalatenduMohanty and @vrutkovs. I addressed all the comments, and, notably, changed the default concurrency to 16.

Any specific reason you changed it to 16 from 8.

File: graph-builder/src/registry.rs
Comment (vrutkovs) [reply graph-builder/src/registry.rs:57]: yep, we no longer fit in 256Mb on stage. Not sure if its due to HashMap or concurrency though

**Comment by steveej** [comment]
> Any specific reason you changed it to 16 from 8.

I wanted to increase the effects and side effects.

**Pull request merged**
